conans/client/graph/compatibility.py

```diff
@@ -91,21 +91,30 @@ def __init__(self, cache):
 
     def compatibles(self, conanfile):
         compat_infos = []
-        if hasattr(conanfile, "compatibility") and callable(conanfile.compatibility):
+        if hasattr(conanfile, "compatibility"):
             with conanfile_exception_formatter(conanfile, "compatibility"):
                 recipe_compatibles = conanfile.compatibility()
                 compat_infos.extend(self._compatible_infos(conanfile, recipe_compatibles))
 
         plugin_compatibles = self._compatibility(conanfile)
         compat_infos.extend(self._compatible_infos(conanfile, plugin_compatibles))
+        if not compat_infos:
+            return {}
 
         result = OrderedDict()
+        original_info = conanfile.info
         for c in compat_infos:
             conanfile.info = c
+            conanfile.settings = c.settings
+            conanfile.options = c.options
             run_validate_package_id(conanfile)
             pid = c.package_id()
             if pid not in result and not c.invalid:
                 result[pid] = c
+        # Restore the original state
+        conanfile.info = original_info
```

Comment:
shouldn't it go under the `finally` block, so it always gets properly restored (e.g. in case of exception)?

Reply:
If there is an exception, the program will crash anyways, not necessary.